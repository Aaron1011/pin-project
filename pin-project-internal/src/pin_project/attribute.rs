use proc_macro2::{Ident, Span, TokenStream};
use quote::{format_ident, quote, quote_spanned, ToTokens};
use syn::{
    parse::{Parse, ParseStream},
    token::Comma,
    *,
};

use crate::utils::SliceExt;

use super::PIN;

// To generate the correct `Unpin` implementation and the projection methods,
// we need to collect the types of the pinned fields.
// However, since proc-macro-attribute is applied before `#[cfg]` and `#[cfg_attr]` on fields,
// we cannot be collecting field types properly at this timing.
// So instead of generating the `Unpin` implementation and the projection methods here,
// delegate their processing to proc-macro-derive.

pub(super) fn parse_attribute(args: TokenStream, mut item: Item) -> Result<TokenStream> {
    let Args { pinned_drop, unsafe_unpin } = syn::parse2(args)?;

    let drop_impl = match &mut item {
        Item::Struct(ItemStruct { attrs, ident, generics, fields, .. }) => {
            super::validate_struct(ident, fields)?;
            push_pin_attr(attrs, unsafe_unpin)?;
            make_drop_impl(ident, generics, pinned_drop)
        }
        Item::Enum(ItemEnum { attrs, ident, generics, brace_token, variants, .. }) => {
            super::validate_enum(*brace_token, variants)?;
            push_pin_attr(attrs, unsafe_unpin)?;
            make_drop_impl(ident, generics, pinned_drop)
        }
        _ => {
            return Err(error!(
                item,
                "#[pin_project] attribute may only be used on structs or enums"
            ));
        }
    };

    let mut item = item.into_token_stream();
    item.extend(drop_impl);
    Ok(item)
}

#[allow(dead_code)] // https://github.com/rust-lang/rust/issues/56750
struct Args {
    pinned_drop: Option<Span>,
    unsafe_unpin: Option<Span>,
}

impl Parse for Args {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let mut pinned_drop = None;
        let mut unsafe_unpin = None;
        while !input.is_empty() {
            let arg = input.parse::<Ident>()?;
            match &*arg.to_string() {
                "PinnedDrop" => {
                    if pinned_drop.is_some() {
                        return Err(error!(arg, "duplicate `PinnedDrop` argument"));
                    }
                    pinned_drop = Some(arg.span());
                }
                "UnsafeUnpin" => {
                    if unsafe_unpin.is_some() {
                        return Err(error!(arg, "duplicate `UnsafeUnpin` argument"));
                    }
                    unsafe_unpin = Some(arg.span());
                }
                _ => {
                    return Err(error!(
                        arg,
                        "an invalid argument was passed to #[pin_project] attribute"
                    ));
                }
            }

            if !input.is_empty() {
                let _: Comma = input.parse()?;
            }
        }

        Ok(Self { pinned_drop, unsafe_unpin })
    }
}

fn push_pin_attr(attrs: &mut Vec<Attribute>, unsafe_unpin: Option<Span>) -> Result<()> {
    if let Some(attr) = attrs.find(PIN) {
        return Err(error!(
            attr,
            "#[pin] attribute may only be used on fields of structs or variants"
        ));
    }

    attrs.push(syn::parse_quote!(#[derive(::pin_project::__private::__PinProjectInternalDerive)]));
    if let Some(span) = unsafe_unpin {
        // Make the error message highlight `UnsafeUnpin` argument.
        let attr = quote_spanned!(span => #[pin(__unsafe_unpin)]);
        attrs.push(syn::parse_quote!(#attr));
    } else {
        attrs.push(syn::parse_quote!(#[pin(__auto_impl_unpin)]));
    }
    Ok(())
}

/// Creates `Drop` implementation for original type.
fn make_drop_impl(ident: &Ident, generics: &Generics, pinned_drop: Option<Span>) -> TokenStream {
    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();

    if let Some(pinned_drop) = pinned_drop {
        // Make the error message highlight `PinnedDrop` argument.
        // See https://github.com/taiki-e/pin-project/issues/16#issuecomment-513586812
        // for why this is only for the span of function calls, not the entire `impl` block.
        let call_drop = quote_spanned! { pinned_drop =>
            ::pin_project::__private::PinnedDrop::drop(pinned_self)
        };

        quote! {
            #[allow(single_use_lifetimes)]
            impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                fn drop(&mut self) {
                    // Safety - we're in 'drop', so we know that 'self' will
                    // never move again.
                    let pinned_self = unsafe { ::core::pin::Pin::new_unchecked(self) };
                    // We call `pinned_drop` only once. Since `PinnedDrop::drop`
                    // is an unsafe function and a private API, it is never called again in safe
                    // code *unless the user uses a maliciously crafted macro*.
                    unsafe {
                        #call_drop;
                    }
                }
            }
        }
    } else {
        // If the user does not provide a pinned_drop impl,
        // we need to ensure that they don't provide a `Drop` impl of their
        // own.
        // Based on https://github.com/upsuper/assert-impl/blob/f503255b292ab0ba8d085b657f4065403cfa46eb/src/lib.rs#L80-L87
        //
        // We create a new identifier for each struct, so that the traits
        // for different types do not conflcit with each other.
        //
        // Another approach would be to provide an empty Drop impl,
        // which would conflict with a user-provided Drop impl.
        // However, this would trigger the compiler's special handling
        // of Drop types (e.g. fields cannot be moved out of a Drop type).
        // This approach prevents the creation of needless Drop impls,
        // giving users more flexibility.
        let trait_ident = format_ident!("{}MustNotImplDrop", ident);

        quote! {
            // There are two possible cases:
            // 1. The user type does not implement Drop. In this case,
            // the first blanked impl will not apply to it. This code
            // will compile, as there is only one impl of MustNotImplDrop for the user type
            // 2. The user type does impl Drop. This will make the blanket impl applicable,
            // which will then comflict with the explicit MustNotImplDrop impl below.
            // This will result in a compilation error, which is exactly what we want.
            trait #trait_ident {}
            #[allow(clippy::drop_bounds)]
            impl<T: ::core::ops::Drop> #trait_ident for T {}
            #[allow(single_use_lifetimes)]
            impl #impl_generics #trait_ident for #ident #ty_generics #where_clause {}
        }
    }
}
