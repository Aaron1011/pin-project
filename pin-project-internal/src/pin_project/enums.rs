use proc_macro2::TokenStream;
use quote::{format_ident, quote, ToTokens};
use syn::{parse::Nothing, Field, Fields, FieldsNamed, FieldsUnnamed, ItemEnum, Result, Variant};

use crate::utils::VecExt;

use super::{Context, PIN};

pub(super) fn parse(cx: &mut Context, mut item: ItemEnum) -> Result<TokenStream> {
    if item.variants.is_empty() {
        return Err(error!(
            item,
            "#[pin_project] attribute may not be used on enums without variants"
        ));
    }
    let has_field = item.variants.iter().try_fold(false, |has_field, v| {
        if let Some((_, e)) = &v.discriminant {
            Err(error!(e, "#[pin_project] attribute may not be used on enums with discriminants"))
        } else if let Fields::Unit = v.fields {
            Ok(has_field)
        } else {
            Ok(true)
        }
    })?;
    if !has_field {
        return Err(error!(
            item.variants,
            "#[pin_project] attribute may not be used on enums that have no field"
        ));
    }

    let (proj_variants, proj_arms) = variants(cx, &mut item)?;

    let proj_ident = &cx.proj_ident;
    let proj_generics = cx.proj_generics();
    let where_clause = item.generics.split_for_impl().2;

    let mut proj_items = quote! {
        #[allow(clippy::mut_mut)] // This lint warns `&mut &mut <ty>`.
        #[allow(dead_code)] // This lint warns unused fields/variants.
        enum #proj_ident #proj_generics #where_clause { #(#proj_variants,)* }
    };

    let project_body = quote! {
        unsafe {
            match self.as_mut().get_unchecked_mut() {
                #(#proj_arms,)*
            }
        }
    };

    let project_into_body = quote! {
        unsafe {
            match self.get_unchecked_mut() {
                #(#proj_arms,)*
            }
        }
    };

    proj_items.extend(cx.make_trait_impl(&project_body, &project_into_body));

    let mut item = item.into_token_stream();
    item.extend(proj_items);
    Ok(item)
}

fn variants(cx: &mut Context, item: &mut ItemEnum) -> Result<(Vec<TokenStream>, Vec<TokenStream>)> {
    let mut proj_variants = Vec::with_capacity(item.variants.len());
    let mut proj_arms = Vec::with_capacity(item.variants.len());
    for Variant { fields, ident, .. } in &mut item.variants {
        let (proj_pat, proj_body, proj_fields) = match fields {
            Fields::Unnamed(fields) => unnamed(cx, fields)?,
            Fields::Named(fields) => named(cx, fields)?,
            Fields::Unit => (TokenStream::new(), TokenStream::new(), TokenStream::new()),
        };
        let Context { orig_ident, proj_ident, .. } = &cx;
        let proj_arm = quote!(#orig_ident::#ident #proj_pat => #proj_ident::#ident #proj_body );
        let proj_variant = quote!(#ident #proj_fields);
        proj_arms.push(proj_arm);
        proj_variants.push(proj_variant);
    }

    Ok((proj_variants, proj_arms))
}

fn named(
    cx: &mut Context,
    FieldsNamed { named: fields, .. }: &mut FieldsNamed,
) -> Result<(TokenStream, TokenStream, TokenStream)> {
    let mut proj_pat = Vec::with_capacity(fields.len());
    let mut proj_body = Vec::with_capacity(fields.len());
    let mut proj_fields = Vec::with_capacity(fields.len());
    for Field { attrs, ident, ty, .. } in fields {
        if let Some(attr) = attrs.find_remove(PIN) {
            let _: Nothing = syn::parse2(attr.tokens)?;
            cx.push_unpin_bounds(ty.clone());
            let lifetime = &cx.lifetime;
            proj_fields.push(quote!(#ident: ::core::pin::Pin<&#lifetime mut #ty>));
            proj_body.push(quote!(#ident: ::core::pin::Pin::new_unchecked(#ident)));
        } else {
            let lifetime = &cx.lifetime;
            proj_fields.push(quote!(#ident: &#lifetime mut #ty));
            proj_body.push(quote!(#ident));
        }
        proj_pat.push(ident);
    }

    let proj_pat = quote!({ #(#proj_pat),* });
    let proj_body = quote!({ #(#proj_body),* });
    let proj_fields = quote!({ #(#proj_fields),* });
    Ok((proj_pat, proj_body, proj_fields))
}

fn unnamed(
    cx: &mut Context,
    FieldsUnnamed { unnamed: fields, .. }: &mut FieldsUnnamed,
) -> Result<(TokenStream, TokenStream, TokenStream)> {
    let mut proj_pat = Vec::with_capacity(fields.len());
    let mut proj_body = Vec::with_capacity(fields.len());
    let mut proj_fields = Vec::with_capacity(fields.len());
    for (i, Field { attrs, ty, .. }) in fields.iter_mut().enumerate() {
        let x = format_ident!("_x{}", i);
        if let Some(attr) = attrs.find_remove(PIN) {
            let _: Nothing = syn::parse2(attr.tokens)?;
            cx.push_unpin_bounds(ty.clone());
            let lifetime = &cx.lifetime;
            proj_fields.push(quote!(::core::pin::Pin<&#lifetime mut #ty>));
            proj_body.push(quote!(::core::pin::Pin::new_unchecked(#x)));
        } else {
            let lifetime = &cx.lifetime;
            proj_fields.push(quote!(&#lifetime mut #ty));
            proj_body.push(quote!(#x));
        }
        proj_pat.push(x);
    }

    let proj_pat = quote!((#(#proj_pat),*));
    let proj_body = quote!((#(#proj_body),*));
    let proj_fields = quote!((#(#proj_fields),*));
    Ok((proj_pat, proj_body, proj_fields))
}
