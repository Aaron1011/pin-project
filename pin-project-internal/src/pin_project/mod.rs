use proc_macro2::{Ident, Span, TokenStream};
use quote::{quote, quote_spanned};
use syn::{
    parse::{Parse, ParseStream},
    Fields, FieldsNamed, FieldsUnnamed, Generics, Index, Item, ItemStruct, Meta, NestedMeta,
    Result, Type,
};

use crate::utils::crate_path;

mod enums;
mod structs;

/// The annotation for pinned type.
const PIN: &str = "pin";

pub(super) fn attribute(args: TokenStream, input: TokenStream) -> TokenStream {
    parse(args, input).unwrap_or_else(|e| e.to_compile_error())
}

#[derive(Clone, Copy)]
struct Args {
    pinned_drop: Option<Span>,
    unsafe_unpin: Option<Span>,
}

impl Args {
    fn impl_drop(self, generics: &Generics) -> ImplDrop<'_> {
        ImplDrop::new(generics, self.pinned_drop)
    }

    fn impl_unpin(self, generics: &Generics) -> ImplUnpin {
        ImplUnpin::new(generics, self.unsafe_unpin)
    }
}

impl Parse for Args {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let mut pinned_drop = None;
        let mut unsafe_unpin = None;
        while !input.is_empty() {
            let i = input.parse::<Ident>()?;
            match &*i.to_string() {
                "PinnedDrop" => pinned_drop = Some(i.span()),
                "UnsafeUnpin" => unsafe_unpin = Some(i.span()),
                _ => return Err(error!(i, "an invalid argument was passed")),
            }
        }
        Ok(Self { pinned_drop, unsafe_unpin })
    }
}

fn parse(args: TokenStream, input: TokenStream) -> Result<TokenStream> {
    let args = syn::parse2(args)?;
    match syn::parse2(input)? {
        Item::Struct(item) => {
            let packed_check = ensure_not_packed(&item)?;
            let mut res = structs::parse(args, item)?;
            res.extend(packed_check);
            Ok(res)
        }
        Item::Enum(item) => {
            // We don't need to check for '#[repr(packed)]',
            // since it does not apply to enums
            enums::parse(args, item)
        }
        item => Err(error!(item, "may only be used on structs or enums")),
    }
}

fn ensure_not_packed(item: &ItemStruct) -> Result<TokenStream> {
    for meta in item.attrs.iter().filter_map(|attr| attr.parse_meta().ok()) {
        if let Meta::List(l) = meta {
            if l.ident == "repr" {
                for repr in l.nested.iter() {
                    if let NestedMeta::Meta(Meta::Word(w)) = repr {
                        if w == "packed" {
                            return Err(error!(
                                w,
                                "pin_project may not be used on #[repr(packed)] types"
                            ));
                        }
                    }
                }
            }
        }
    }

    // Workaround for https://github.com/taiki-e/pin-project/issues/32
    // Through the tricky use of proc macros, it's possible to bypass
    // the above check for the 'repr' attribute.
    // To ensure that it's impossible to use pin projections on a #[repr(packed)][
    // struct, we generate code like this:
    //
    // #[deny(safe_packed_borrows)]
    // fn enforce_not_packed_for_MyStruct(val: MyStruct) {
    //  let _field1 = &val.field1;
    //  let _field2 = &val.field2;
    //  ...
    //  let _fieldn = &val.fieldn;
    // }
    //
    // Taking a reference to a packed field is unsafe, amd appplying
    // #[deny(safe_packed_borrows)] makes sure that doing this without
    // an 'unsafe' block (which we deliberately do not generate)
    // is a hard error.
    //
    // If the struct ends up having #[repr(packed)] applied somehow,
    // this will generate an (unfriendly) error message. Under all reasonable
    // circumstances, we'll detect the #[repr(packed)] attribute, and generate
    // a much nicer error above.
    //
    // There is one exception: If the type of a struct field has a alignemtn of 1
    // (e.g. u8), it is always safe to take a reference to it, even if the struct
    // is #[repr(packed)]. If the struct is composed entirely of types of alignent 1,
    // our generated method will not trigger an error if the struct is #[repr(packed)]
    //
    // Fortunately, this should have no observable consequence - #[repr(packed)]
    // is essentially a no-op on such a type. Nevertheless, we include a test
    // to ensure that the compiler doesn't ever try to copy the fields on
    // such a struct when trying to drop it - which is reason we prevent
    // #[repr(packed)] in the first place
    let mut field_refs = vec![];
    match &item.fields {
        Fields::Named(FieldsNamed { named, .. }) => {
            for field in named.iter() {
                let ident = field.ident.as_ref().unwrap();
                field_refs.push(quote!(&val.#ident;));
            }
        }
        Fields::Unnamed(FieldsUnnamed { unnamed, .. }) => {
            for (i, _) in unnamed.iter().enumerate() {
                let index = Index::from(i);
                field_refs.push(quote!(&val.#index;));
            }
        }
        Fields::Unit => {}
    }

    let (impl_generics, ty_generics, where_clause) = item.generics.split_for_impl();

    let struct_name = &item.ident;
    let method_name = Ident::new(
        &("__pin_project_assert_not_repr_packed_".to_string() + &item.ident.to_string()),
        Span::call_site(),
    );
    let test_fn = quote! {
        #[deny(safe_packed_borrows)]
        fn #method_name #impl_generics (val: #struct_name #ty_generics) #where_clause {
            #(#field_refs)*
        }
    };
    Ok(test_fn)
}

/// Makes the generics of projected type from the reference of the original generics.
fn proj_generics(generics: &Generics) -> Generics {
    let mut generics = generics.clone();
    generics.params.insert(0, syn::parse_quote!('__a));
    generics
}

// =================================================================================================
// Drop implementation

struct ImplDrop<'a> {
    generics: &'a Generics,
    pinned_drop: Option<Span>,
}

impl<'a> ImplDrop<'a> {
    fn new(generics: &'a Generics, pinned_drop: Option<Span>) -> Self {
        Self { generics, pinned_drop }
    }

    fn build(self, ident: &Ident) -> TokenStream {
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();

        if let Some(pinned_drop) = self.pinned_drop {
            let crate_path = crate_path();
            let call = quote_spanned! { pinned_drop =>
                ::#crate_path::__private::UnsafePinnedDrop::pinned_drop(pinned_self)
            };

            quote! {
                impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                    fn drop(&mut self) {
                        // Safety - we're in 'drop', so we know that 'self' will
                        // never move again
                        let pinned_self = unsafe { ::core::pin::Pin::new_unchecked(self) };
                        // We call `pinned_drop` only once. Since `UnsafePinnedDrop::pinned_drop`
                        // is an unsafe function and a private API, it is never called again in safe
                        // code *unless the user uses a maliciously crafted macro*.
                        unsafe {
                            #call;
                        }
                    }
                }
            }
        } else {
            quote! {
                impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                    fn drop(&mut self) {
                        // Do nothing. The precense of this Drop
                        // impl ensures that the user can't provide one of their own
                    }
                }
            }
        }
    }
}

// =================================================================================================
// conditional Unpin implementation

struct ImplUnpin {
    generics: Generics,
    unsafe_unpin: bool,
}

impl ImplUnpin {
    fn new(generics: &Generics, unsafe_unpin: Option<Span>) -> Self {
        let mut generics = generics.clone();
        if let Some(unsafe_unpin) = unsafe_unpin {
            let crate_path = crate_path();
            generics.make_where_clause().predicates.push(
                syn::parse2(quote_spanned! { unsafe_unpin =>
                    ::#crate_path::__private::Wrapper<Self>: ::#crate_path::UnsafeUnpin
                })
                .unwrap(),
            );
        }

        Self { generics, unsafe_unpin: unsafe_unpin.is_some() }
    }

    fn push(&mut self, ty: &Type) {
        // We only add bounds for automatically generated impls
        if !self.unsafe_unpin {
            self.generics
                .make_where_clause()
                .predicates
                .push(syn::parse_quote!(#ty: ::core::marker::Unpin));
        }
    }

    /// Creates `Unpin` implementation.
    fn build(self, ident: &Ident) -> TokenStream {
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
        quote! {
            impl #impl_generics ::core::marker::Unpin for #ident #ty_generics #where_clause {}
        }
    }
}
