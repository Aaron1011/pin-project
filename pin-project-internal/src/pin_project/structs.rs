use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use syn::{parse::Nothing, Field, Fields, FieldsNamed, FieldsUnnamed, Index, ItemStruct, Result};

use crate::utils::VecExt;

use super::{Context, PIN};

pub(super) fn parse(cx: &mut Context, mut item: ItemStruct) -> Result<TokenStream> {
    let (proj_fields, proj_init) = match &mut item.fields {
        Fields::Named(FieldsNamed { named: fields, .. })
        | Fields::Unnamed(FieldsUnnamed { unnamed: fields, .. })
            if fields.is_empty() =>
        {
            return Err(error!(item.fields, "cannot be implemented for structs with zero fields"))
        }
        Fields::Unit => return Err(error!(item, "cannot be implemented for structs with units")),

        Fields::Named(fields) => named(cx, fields)?,
        Fields::Unnamed(fields) => unnamed(cx, fields)?,
    };

    let Context { proj_ident, proj_trait, orig_ident, lifetime, .. } = &cx;
    let proj_generics = cx.proj_generics();
    let proj_ty_generics = proj_generics.split_for_impl().1;
    let (impl_generics, ty_generics, where_clause) = item.generics.split_for_impl();

    let mut proj_items = quote! {
        struct #proj_ident #proj_generics #where_clause #proj_fields
    };
    proj_items.extend(quote! {
        impl #impl_generics #proj_trait #ty_generics for ::core::pin::Pin<&mut #orig_ident #ty_generics> #where_clause {
            fn project<#lifetime>(&#lifetime mut self) -> #proj_ident #proj_ty_generics #where_clause {
                unsafe {
                    let this = self.as_mut().get_unchecked_mut();
                    #proj_ident #proj_init
                }
            }
        }
    });

    let mut item = item.into_token_stream();
    item.extend(proj_items);
    Ok(item)
}

fn named(
    cx: &mut Context,
    FieldsNamed { named: fields, .. }: &mut FieldsNamed,
) -> Result<(TokenStream, TokenStream)> {
    let mut proj_fields = Vec::with_capacity(fields.len());
    let mut proj_init = Vec::with_capacity(fields.len());
    for Field { attrs, ident, ty, .. } in fields {
        if let Some(attr) = attrs.find_remove(PIN) {
            let _: Nothing = syn::parse2(attr.tokens)?;
            cx.push_unpin_bounds(ty);
            let lifetime = &cx.lifetime;
            proj_fields.push(quote!(#ident: ::core::pin::Pin<&#lifetime mut #ty>));
            proj_init.push(quote!(#ident: ::core::pin::Pin::new_unchecked(&mut this.#ident)));
        } else {
            let lifetime = &cx.lifetime;
            proj_fields.push(quote!(#ident: &#lifetime mut #ty));
            proj_init.push(quote!(#ident: &mut this.#ident));
        }
    }

    let proj_fields = quote!({ #(#proj_fields,)* });
    let proj_init = quote!({ #(#proj_init,)* });
    Ok((proj_fields, proj_init))
}

fn unnamed(
    cx: &mut Context,
    FieldsUnnamed { unnamed: fields, .. }: &mut FieldsUnnamed,
) -> Result<(TokenStream, TokenStream)> {
    let mut proj_fields = Vec::with_capacity(fields.len());
    let mut proj_init = Vec::with_capacity(fields.len());
    for (i, Field { attrs, ty, .. }) in fields.iter_mut().enumerate() {
        let i = Index::from(i);
        if let Some(attr) = attrs.find_remove(PIN) {
            let _: Nothing = syn::parse2(attr.tokens)?;
            cx.push_unpin_bounds(ty);
            let lifetime = &cx.lifetime;
            proj_fields.push(quote!(::core::pin::Pin<&#lifetime mut #ty>));
            proj_init.push(quote!(::core::pin::Pin::new_unchecked(&mut this.#i)));
        } else {
            let lifetime = &cx.lifetime;
            proj_fields.push(quote!(&#lifetime mut #ty));
            proj_init.push(quote!(&mut this.#i));
        }
    }

    let proj_fields = quote!((#(#proj_fields,)*););
    let proj_init = quote!((#(#proj_init,)*));
    Ok((proj_fields, proj_init))
}
