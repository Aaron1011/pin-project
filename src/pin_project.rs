use proc_macro2::{Group, Ident, TokenStream, TokenTree};
use quote::ToTokens;
use syn::{
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    visit_mut::VisitMut,
    *,
};

use crate::project::Dummy;

pub(super) fn attribute(args: TokenStream, input: TokenStream) -> TokenStream {
    syn::parse2(input)
        .and_then(|mut item| {
            syn::parse2(args).map(|args: Args| {
                // TODO: Integrate into `replace_item_fn`?.
                Dummy.visit_item_fn_mut(&mut item);
                replace_item_fn(&args.0, &mut item);
                item.into_token_stream()
            })
        })
        .unwrap_or_else(|e| e.to_compile_error())
}

fn replace_item_fn(args: &[Ident], ItemFn { decl, block, .. }: &mut ItemFn) {
    decl.inputs.iter_mut().for_each(|input| match input {
        FnArg::Captured(ArgCaptured {
            pat: Pat::Ident(pat @ PatIdent { subpat: None, .. }),
            ..
        }) if args.contains(&pat.ident) => {
            let mut local = Local {
                attrs: Vec::new(),
                let_token: token::Let::default(),
                pats: Punctuated::new(),
                ty: None,
                init: None,
                semi_token: token::Semi::default(),
            };
            let (local_pat, init) = if pat.ident == "self" {
                ReplaceSelf.visit_block_mut(block);
                let mut local_pat = pat.clone();
                prepend_underscores_to_self(&mut local_pat.ident);
                (local_pat, syn::parse_quote!(self.project()))
            } else {
                let ident = &pat.ident;
                (pat.clone(), syn::parse_quote!(#ident.project()))
            };
            local.pats.push(Pat::Ident(local_pat));
            local.init = Some((token::Eq::default(), init));
            block.stmts.insert(0, Stmt::Local(local));

            if pat.by_ref.is_none() {
                pat.mutability = None;
            }
            pat.by_ref = None;
        }
        _ => {}
    })
}

struct Args(Vec<Ident>);

impl Parse for Args {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        let mut args = Vec::new();
        let mut first = true;
        while !input.is_empty() {
            if first {
                first = false;
            } else {
                let _: Token![,] = input.parse()?;
                if input.is_empty() {
                    break;
                }
            }

            let ident = if input.peek(Token![self]) {
                let t: Token![self] = input.parse()?;
                Ident::new("self", t.span)
            } else {
                input.parse()?
            };
            if args.contains(&ident) {
                // TODO: error
            } else {
                args.push(ident);
            }
        }
        Ok(Self(args))
    }
}

// https://github.com/dtolnay/no-panic/blob/master/src/lib.rs

struct ReplaceSelf;

impl VisitMut for ReplaceSelf {
    fn visit_expr_path_mut(&mut self, i: &mut ExprPath) {
        if i.qself.is_none() && i.path.is_ident("self") {
            prepend_underscores_to_self(&mut i.path.segments[0].ident);
        }
    }

    fn visit_macro_mut(&mut self, i: &mut Macro) {
        // We can't tell in general whether `self` inside a macro invocation
        // refers to the self in the argument list or a different self
        // introduced within the macro. Heuristic: if the macro input contains
        // `fn`, then `self` is more likely to refer to something other than the
        // outer function's self argument.
        if !contains_fn(i.tts.clone()) {
            i.tts = fold_token_stream(i.tts.clone());
        }
    }

    fn visit_item_mut(&mut self, _i: &mut Item) {
        // Do nothing, as `self` now means something else.
    }
}

fn contains_fn(tts: TokenStream) -> bool {
    tts.into_iter().any(|tt| match tt {
        TokenTree::Ident(ident) => ident == "fn",
        TokenTree::Group(group) => contains_fn(group.stream()),
        _ => false,
    })
}

fn fold_token_stream(tts: TokenStream) -> TokenStream {
    tts.into_iter()
        .map(|tt| match tt {
            TokenTree::Ident(mut ident) => {
                prepend_underscores_to_self(&mut ident);
                TokenTree::Ident(ident)
            }
            TokenTree::Group(group) => {
                let content = fold_token_stream(group.stream());
                TokenTree::Group(Group::new(group.delimiter(), content))
            }
            other => other,
        })
        .collect()
}

fn prepend_underscores_to_self(ident: &mut Ident) {
    if ident == "self" {
        *ident = Ident::new("__self", ident.span());
    }
}
