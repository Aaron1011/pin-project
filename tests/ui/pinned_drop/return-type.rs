// compile-fail

#![deny(warnings)]

use pin_project::pin_project;
use std::pin::Pin;

pin_project! {
    #[pin_projectable]
    pub struct Foo<'a> {
        was_dropped: &'a mut bool,
        #[pin] field_2: u8
    }

    #[pinned_drop]
    fn do_drop(foo: Pin<&mut Foo<'_>>) -> bool { //~ ERROR #[pinned_drop] function must return the unit type
        **foo.project().was_dropped = true;
        true
    }
}

pin_project! {
    #[pin_projectable]
    pub struct Bar<'a> {
        was_dropped: &'a mut bool,
        #[pin] field_2: u8
    }

    #[pinned_drop]
    fn do_drop(foo: Pin<&mut Bar<'_>>) -> () { // OK
        **foo.project().was_dropped = true;
    }
}

fn main() {}
