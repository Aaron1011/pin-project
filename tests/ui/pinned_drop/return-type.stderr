error: #[pinned_drop] function must return the unit type
  --> $DIR/return-type.rs:16:43
   |
16 |     fn do_drop(foo: Pin<&mut Foo<'_>>) -> bool { //~ ERROR #[pinned_drop] function must return the unit type
   |                                           ^^^^

error: aborting due to previous error

